javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt

```diff
@@ -0,0 +1,262 @@
+            expectHandleException(writeHandleName) {
+                handleHolder.writeHandle
+            }
+
+            val writeHandle = hm.createSingletonHandle(
+                handleHolder,
+                writeHandleName,
+                singletonKey,
+                schema,
+                HandleMode.Write
+            )
+            assertThat(writeHandle).isInstanceOf(WritableSingleton::class.java)
+            assertThat(writeHandle).isNotInstanceOf(ReadableSingleton::class.java)
+            handleHolder.writeHandle.store(entity1)
+
+            // Now read back from a different handle
+            val readHandleName = "readHandle"
+
+            expectHandleException(readHandleName) {
+                handleHolder.readHandle
+            }
+
+            val readHandle = hm.createSingletonHandle(
+                handleHolder,
+                readHandleName,
+                singletonKey,
+                schema,
+                HandleMode.Read
+            )
+            val readBack = handleHolder.readHandle.fetch()
+            assertThat(readHandle).isInstanceOf(ReadableSingleton::class.java)
+            assertThat(readHandle).isNotInstanceOf(WritableSingleton::class.java)
+
+            assertThat(readBack).isEqualTo(entity1)
+
+            // Now read back from a different handle
+            val readWriteHandleName = "readWriteHandle"
+
+            expectHandleException(readWriteHandleName) {
+                handleHolder.readWriteHandle
+            }
+
+            val readWriteHandle = hm.createSingletonHandle(
+                handleHolder,
+                readWriteHandleName,
+                singletonKey,
+                schema,
+                HandleMode.ReadWrite
+            )
+            val readBack2 = handleHolder.readWriteHandle.fetch()
+            assertThat(readWriteHandle).isInstanceOf(ReadWriteSingleton::class.java)
+            assertThat(readBack2).isEqualTo(entity1)
+
+            val updatedEntity: Person? = suspendCoroutine { continuation ->
+                // Verify callbacks work
+                launch {
+                    handleHolder.readWriteHandle.onUpdate {
+                        continuation.resume(it)
+                    }
+                    handleHolder.writeHandle.store(entity2)
+                }
+            }
+
+            assertThat(updatedEntity).isEqualTo(entity2)
+        }
+    }
+
+    @Test
+    fun setHandle_writeFollowedByReadWithOnUpdate() = runBlockingTest {
+        handleManagerTest { hm ->
+            val handleHolder = TestParticleHandles()
+            val writeSetHandleName = "writeSetHandle"
+
+            expectHandleException(writeSetHandleName) {
+                handleHolder.writeSetHandle
+            }
+
+            val writeSetHandle = hm.createSetHandle(
+                handleHolder,
+                writeSetHandleName,
+                setKey,
+                schema,
+                HandleMode.Write
+            )
```

Comment:
Optional: might be more readable to create a helper method for these. It looks like the arguments rarely change.

```
private fun createSetHandle(
    handleManager: HandleManager,
    handleName: String,
    handleMode: HandleMode
) = handleManager.createSetHandle(
    handleHolder,
    handleName,
    setKey,
    schema,
    handleMode
)
```

(`handleHolder` could be a member variable and be initialised in `setUp`)

Reply:
Done